#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: getdents64 system call is failed before linux kernel
# DP:	2.4.0-test7. This patch solves getdents64() failure on kernel 2.2,
# DP:	and it tries to call getdents().
# DP: Author: GOTO Masanori <gotom@debian.org>
# DP: Upstream status: Submitted
# DP: Status Details: Applied already in CVS
# DP: Date: 2002-12-23

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p0 < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p0 < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
--- sysdeps/unix/sysv/linux/getdents.c.org	2002-12-14 02:09:29.000000000 +0900
+++ sysdeps/unix/sysv/linux/getdents.c	2002-12-23 02:15:12.000000000 +0900
@@ -126,7 +126,7 @@
       retval = INLINE_SYSCALL (getdents64, 3, fd, CHECK_N(kbuf, kbytes),
 			       kbytes);
 # ifndef __ASSUME_GETDENTS64_SYSCALL
-      if (retval != -1 || errno != EINVAL)
+      if (retval != -1 || (errno != EINVAL && errno != ENOSYS))
 # endif
 	{
 	  const size_t size_diff = (offsetof (struct kernel_dirent64, d_name)
